Substats M1

heipowerman opened the pull request:
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/LSRr7PCjBpEbKGh89) has been filled out for this milestone.
- [x] This pull request is being made by the same account as the accepted application.
- [x] In case of acceptance, the payment will be transferred to the BTC/ETH payment address provided in the application.
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/1153

semuelle commented on deliveries/Substats-Milestone_1.md:
```diff
@@ -0,0 +1,16 @@
+
+**The [invoice form :pencil:](https://docs.google.com/forms/d/e/1FAIpQLSfmNYaoCgrxyhzgoKQ0ynQvnNRoTmgApz9NrMp-hd8mhIiO0A/viewform) has been filled out correctly for this milestone and the delivery is according to the official [milestone delivery guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).**  
+
+* **Application Document:** https://github.com/CESSProject/Grants-Program/blob/master/applications/substats.md  
```
```suggestion
* **Application Document:** https://github.com/w3f/Grants-Program/blob/master/applications/substats.md  
```

heipowerman replied:
Thanks for your remind !

dsm-w3f commented:
@heipowerman  thank you for your delivery. I'm trying to spin up all system services to evaluate the delivery. However, I notice that no Docker or docker-compose file were provided to facilitate this task. Would be nice to have a docker-compose file that config and startup all the system services. I tried to config and connect the system with 8.0.31 MySQL Community Server - GPL and is not working with the error below:

`unhandledRejection Error: ER_NOT_SUPPORTED_AUTH_MODE: Client does not support authentication protocol requested by server; consider upgrading MySQL client`

Can you provide assistance on that? Which MySQL version is supported? Or even better, could you provide a docker-compose file that start-up all system services?

dsm-w3f commented:
@heipowerman thank you for the improvements. I was able to start the system using docker, however some test failed and the API calls seem to have some problems. Please see [the evaluation document](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/substats_1_dsm-w3f.md) for details. Let me know when I can continue this evaluation.

heipowerman commented:
@dsm-w3f Thank you for finding the problem. It seems a problem with the program running environment. About API Testing error ‘api.query.fileBank not a function’, cause the connected chain has not API, so we remove it from [api-docs.md](https://github.com/CESSProject/substats/blob/master/documents/api-docs.md).
About the Automated Tests error, the reason is that Mysql configuration file is not modified. Please try ‘npm start ./mysql-config.json’
Please refer to: https://github.com/CESSProject/substats/blob/master/README.md#1-run-api-server
We also created a database sqlite3 as a backup to run program on systems if mysql is not installed.
At lease, we have updated docker image.

dsm-w3f commented:
@heipowerman thank you for the improvements. Now the tests are passing, however the API is still with the same behavior. See details on [the evaluation document](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/substats_1_dsm-w3f.md). Let me know when I can continue this evaluation.

heipowerman commented:
@dsm-w3f thanks for your reply. we have tried to fix the error, please check it below:

# Evaluation V2

##### 1.About API Testing

Please check Evaluation V1

##### 2.About Automated Tests

Due to time constraints, we only added tests for the core functions.If you have a requirement for unit test coverage, please tell us the percentage that the tests need to reach. In the meanwhile, we will complete unit testing continually.



# Evaluation V1

### API Testing

##### 1. Data Dictionary

> response 
>
> ```javascript
> {
> 	"msg": "ok",
> 	"data": []
> }
> ```
>
> This output is not error, it is common. When initializing the database, the program added default data.Please check:
>
> [https://github.com/CESSProject/substats/blob/0648c7a7644630666a9bb43a671d7a8a51db49ad/documents/database-init.sql#L183](https://github.com/CESSProject/substats/blob/0648c7a7644630666a9bb43a671d7a8a51db49ad/documents/database-init.sql#L183)

##### 2. Chain State Query

>response 
>
>```javascript
>{
>	"msg": "api.query.fileBank not a function"
>}
>```
>
>Because some chains do not have this API (api.query.fileBank),so this api has been removed from the documentation

More details, please check the API doc:
https://github.com/CESSProject/substats/blob/master/documents/api-docs.md


##### 3. Chain Consts

This output is not error, it is common. 

##### 4. Query List and Query Details

> Take a long time and didn't return any result from the API call.

The reason may be that the server port is not open, so the server is not connected

If the port is not open, please open port 8080 on the firewall.

Pleaes confirm that the port is open and the web service is working.Please try this link:

http://【your server ip】:8080/system-status.html 


### Automated Tests

The error is related to the configuration. We have done the compatibility to fix the error.

dsm-w3f commented:
@heipowerman thank you for the answers. After the changes, the substats container was not able to connect to the database due to a password change. I needed to change the demo/mysql-config.json to the password stated in the docker-compose.yml file to be able to connect.

I'm not convinced that the system tests (manual API tests) are ok for passing this evaluation. There is little or no data in the database for testing. The API is still returning empty data or no response errors. The ports are open and I'm able to check the system status as shown in the print below. Please improve the data/system for API system test in order to properly show that it works.

![image](https://user-images.githubusercontent.com/112647953/209943055-27db5e6d-eb7f-4125-a720-2e4b13f4bb01.png)

Regarding the automated tests, there is no restriction of coverage for automated testing. Besides that, we always evaluate to see if the system was properly tested. Remember that all data of the evaluation is public, and usually, users/stakeholders expect high-quality systems.

heipowerman commented:
@dsm-w3f Thanks for your reply. we have tried to fix the problem:
  1  When the system was started, we added a program to synchronize block information from the chain to the database. When you test the API service, the data will be returned. The entry code is here: https://github.com/CESSProject/substats/blob/a20719f77624a3f2658a3562cf041192500b7a89/app.js#L93

  2  When docker is running, we found that some server is failed to connect MySQL, we treid to improve the system. We used the sqlite database in docker, which will make the project run in an independent sandbox environment without request external MySQL. Please check README: https://github.com/CESSProject/substats/blob/master/README.md?plain=1#L115

  3  About the automated testing seems few, we have added the unit test code and improved the coverage of unit testing.

  4  We also fixed MySQL password error in demo/mysql-config.json https://github.com/CESSProject/substats/blob/a20719f77624a3f2658a3562cf041192500b7a89/demo/mysql-config.json#L5

dsm-w3f commented:
@heipowerman thank you for the improvements. The milestone delivery was approved. I'll forward your invoice internally and the payment should take place within two weeks (could happen before). As stated in [the evaluation document](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/substats_1_dsm-w3f.md), just make sure that all test cases have examples with returning data in the next milestone delivery. Good work!

heipowerman commented:
@dsm-w3f  Thanks for patiently comment and reply. When we submit the code next time, we will check the code carefully

fededubbi commented:
Hi all,

I noticed you sent us an invoice with the same number of a previous one.
Could you please resend it with the new number?

Many thanks,
Federica

RouvenP commented:
hi @heipowerman we transferred the payment today.
